pkg/image/prune/orphanedimagefinder.go

```diff
@@ -0,0 +1,162 @@
+package prune
+
+import (
+	"strings"
+
+	kapi "github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+	"github.com/golang/glog"
+	gonum "github.com/gonum/graph"
+	"github.com/openshift/origin/pkg/api/graph"
+	imageapi "github.com/openshift/origin/pkg/image/api"
+)
+
+type OrphanedImageFinder interface {
+	Find() []*imageapi.Image
+}
+
+type orphanedImageFinder struct {
+	g graph.Graph
+}
+
+var _ OrphanedImageFinder = &orphanedImageFinder{}
+
+func NewOrphanedImageFinder(registryURL string, images *imageapi.ImageList, streams *imageapi.ImageStreamList, pods *kapi.PodList, rcs *kapi.ReplicationControllerList) OrphanedImageFinder {
+	g := graph.New()
+
+	addImagesToGraph(g, images, registryURL)
+	addImageStreamsToGraph(g, streams, registryURL)
+	addPodsToGraph(g, pods, registryURL)
+	addReplicationControllersToGraph(g, rcs, registryURL)
+
+	return &orphanedImageFinder{g: g}
+}
+
+const minimumAgeInMinutesToPrune = 60
+
+func addImagesToGraph(g graph.Graph, images *imageapi.ImageList, registryURL string) {
+	for i := range images.Items {
+		image := &images.Items[i]
+		if !strings.HasPrefix(image.DockerImageReference, registryURL) {
+			continue
+		}
+		age := util.Now().Sub(image.CreationTimestamp.Time)
+		if age.Minutes() < minimumAgeInMinutesToPrune {
+			continue
+		}
+		graph.Image(g, image)
+	}
+}
+
+func addImageStreamsToGraph(g graph.Graph, streams *imageapi.ImageStreamList, registryURL string) {
+	for i := range streams.Items {
+		stream := &streams.Items[i]
+		if len(stream.Spec.DockerImageRepository) > 0 {
+			// skip streams that point at external registries
+			continue
+		}
+		imageStreamNode := graph.ImageStream(g, stream)
+		for _, history := range stream.Status.Tags {
+			if len(history.Items) == 0 {
+				continue
+			}
+			for i := range history.Items {
+				glog.Infof("Looking for node with image id %q", history.Items[i].Image)
+				imageNode := graph.FindImage(g, history.Items[i].Image)
+				glog.Infof("Found: %#v", imageNode)
+				if imageNode != nil {
+					var kind int
+					switch i {
+					case 0:
+						kind = graph.ReferencedImageGraphEdgeKind
+					default:
+						kind = graph.WeakReferencedImageGraphEdgeKind
+					}
+					g.AddEdge(imageStreamNode, imageNode, kind)
+				}
+			}
+		}
+	}
+}
+
+func addPodsToGraph(g graph.Graph, pods *kapi.PodList, registryURL string) {
+	for i := range pods.Items {
+		pod := &pods.Items[i]
+
+		age := util.Now().Sub(pod.CreationTimestamp.Time)
+		if age.Minutes() < minimumAgeInMinutesToPrune {
```

Comment:
This is something that should be on a pruneAlgorithm object (the value) - if you do that, you can move registryURL there